s3proxy/internal/router/router.go

```diff
@@ -0,0 +1,416 @@
+	if header.Get("x-amz-copy-source") != "" {
+		return true
+	}
+
+	_, partNumber := query["partNumber"]
+	_, uploadID := query["uploadId"]
+	_, tagging := query["tagging"]
+	_, legalHold := query["legal-hold"]
+	_, objectLock := query["object-lock"]
+	_, retention := query["retention"]
+	_, publicAccessBlock := query["publicAccessBlock"]
+	_, acl := query["acl"]
+
+	return partNumber || uploadID || tagging || legalHold || objectLock || retention || publicAccessBlock || acl
+}
+
+func sha256sum(data []byte) string {
+	digest := sha256.Sum256(data)
+	return fmt.Sprintf("%x", digest)
+}
+
+// getMetadataHeaders parses user-defined metadata headers from a
+// http.Header object. Users can define custom headers by taking
+// HEADERNAME and prefixing it with "x-amz-meta-".
+func getMetadataHeaders(header http.Header) map[string]string {
+	result := map[string]string{}
+
+	for key := range header {
+		key = strings.ToLower(key)
+
+		if strings.HasPrefix(key, "x-amz-meta-") {
+			name := strings.TrimPrefix(key, "x-amz-meta-")
+			result[name] = strings.Join(header.Values(key), ",")
+		}
+	}
+
+	return result
+}
+
+func parseRetentionTime(raw string) (time.Time, error) {
+	if raw == "" {
+		return time.Time{}, nil
+	}
+	return time.Parse(time.RFC3339, raw)
+}
+
+// repackage implements all modifications we need to do to an incoming request that we want to forward to the s3 API.
+func repackage(r *http.Request) http.Request {
+	req := r.Clone(r.Context())
+
+	// HTTP clients are not supposed to set this field, however when we receive a request it is set.
+	// So, we unset it.
+	req.RequestURI = ""
+
+	req.URL.Host = r.Host
+	// We always want to use HTTPS when talking to S3.
+	req.URL.Scheme = "https"
+
+	return *req
+}
+
+// validateContentMD5 checks if the content-md5 header matches the body.
+func validateContentMD5(contentMD5 string, body []byte) error {
+	if contentMD5 == "" {
+		return nil
+	}
+
+	expected, err := base64.StdEncoding.DecodeString(contentMD5)
+	if err != nil {
+		return fmt.Errorf("decoding base64: %w", err)
+	}
+
+	if len(expected) != 16 {
+		return fmt.Errorf("content-md5 must be 16 bytes long, got %d bytes", len(expected))
+	}
+
+	actual := md5.Sum(body)
+
+	if !bytes.Equal(actual[:], expected) {
+		return fmt.Errorf("content-md5 mismatch, header is %x, body is %x", expected, actual)
+	}
+
+	return nil
+}
+
+// match reports whether path matches pattern, and if it matches,
+// assigns any capture groups to the *string or *int vars.
+func match(path, pattern string, vars ...interface{}) bool {
+	regex := mustCompileCached(pattern)
+	matches := regex.FindStringSubmatch(path)
+	if len(matches) <= 0 {
+		return false
+	}
+	for i, match := range matches[1:] {
+		switch p := vars[i].(type) {
+		case *string:
+			*p = match
+		case *int:
+			n, err := strconv.Atoi(match)
+			if err != nil {
+				return false
+			}
+			*p = n
+		default:
+			panic("vars must be *string or *int")
```

Comment:
This function should not panic.
Return errors instead

Reply:
Simplified this so the panic/error is not necessary.